pkg/proxy/netstat.go

```diff
@@ -0,0 +1,52 @@
+package proxy
+
+import (
+	"fmt"
+	"io/ioutil"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	// procNetTCPFiles is the constant list of /proc/net files to read to get
+	// the same information about open TCP connections as output by netstat.
+	procNetTCPFiles = []string{
+		"/proc/net/tcp",
+		"/proc/net/tcp6",
+	}
+
+	// procNetTCPRegexp matches the first two columns of /proc/net/tcp* files
+	// and submatches on the local port number.
+	procNetTCPRegexp = regexp.MustCompile("^ *[[:digit:]]*: *[[:xdigit:]]*:([[:xdigit:]]*) ")
+)
+
+// readOpenLocalTCPPorts returns the set of TCP ports currently open locally.
+func readOpenLocalTCPPorts() (map[uint16]struct{}, error) {
+	openLocalPorts := make(map[uint16]struct{}, 128)
+
+	for _, file := range procNetTCPFiles {
+		bytes, err := ioutil.ReadFile(file)
+		if err != nil {
+			return nil, fmt.Errorf("cannot read proc file %s: %s", file, err)
+		}
+		lines := strings.Split(string(bytes), "\n")
+
+		// Extract the local port number from the "local_address" column.
+		// The header line won't match and will be ignored.
+		for _, line := range lines {
+			groups := procNetTCPRegexp.FindStringSubmatch(line)
+			if len(groups) != 2 { // no match
+				continue
+			}
+			// The port number is in hexadecimal.
+			localPort, err := strconv.ParseUint(groups[1], 16, 16)
+			if err != nil {
+				return nil, fmt.Errorf("invalid local port number in %s: %s", file, err)
```

Comment:
On the contrary, it may be better to even `log.Fatalf()` in that case?

Reply:
Why? It's a single port out of many. I think `log.Fatalf` is extreme for this

Reply:
This error is pretty extreme. It would mean that we parse the line OK, but the port is not in hexadecimal and/or doesn't fit in a 16-bit unsigned int.